src/plugin.ts

```diff
@@ -18,16 +18,32 @@ async function main() {
   const typeMap = createTypeMap(request, options);
   const utils = makeUtils(options);
   const ctx: Context = { typeMap, options, utils };
+  let filesToGenerate;
+
+  if (options.emitImportedFiles) {
+    const fileSet = new Set();
+    const addFilesDeep = (filenames: string[]) => {
+      filenames = filenames.filter((name) => !options.M[name]);
+      filenames.forEach((name) => {
+        fileSet.add(name);
+        const file = request.protoFile.find((file) => file.name === name);
+        if (file && file.dependency.length > 0) {
+          addFilesDeep(file.dependency);
+        }
+      });
+    };
+    addFilesDeep(request.fileToGenerate);
+    filesToGenerate = request.protoFile.filter((file) => fileSet.has(file.name));
+  } else {
+    filesToGenerate = protoFilesToGenerate(request);
+  }
 
-  const filesToGenerate = options.emitImportedFiles ? request.protoFile : protoFilesToGenerate(request);
   const files = await Promise.all(
-    filesToGenerate
-      .filter((file) => !options.M[file.name])
```

Comment:
Won't removing this line mean that with `emitImportedFiles=false`, we are not filtering the `protoFilesToGenerate` against `options.M` anymore?

Or are you thinking that is necessary, b/c `protoFilesToGenerate` already has only the files directly from the CLI, so should be okay w/o this filter?

Reply:
To me, it makes sense that it would be an error to map a file that is explicitly being passed as input to `protoc`. Emitting this error could be a separate PR. So it was intentional not to add filtering in the `emitImportedFiles=false` case - maybe I should add it anyway, to keep previous odd behavior?

Reply:
I added it back. Thanks!